Fix precedencegroup and operator decl lookup

hamishknight opened the pull request:
Re-implement operator and `precedencegroup` decl lookup to use `namelookup::getAllImports` and existing decl shadowing logic. This allows us to find operator decls through `@_exported` imports, prefer operator decls defined in the same module over imported decls, and fixes a couple of other subtle issues.

Because this new implementation is technically source breaking, as we can find multiple results where we used to only find one result, it's placed behind a Frontend staging flag `-enable-new-operator-lookup` (with the aim of enabling it by default when we get a new language mode).

However the new logic will always be used if the result is unambiguous. This means that e.g `@_exported` operators will be instantly available as long as there's only one candidate. If multiple candidates are found, we fall back to the old logic.

Resolves SR-12132.
Resolves rdar://48731166 & rdar://59198796.

hamishknight commented:
https://github.com/apple/llvm-project/pull/892

@swift-ci please test

hamishknight commented:
@swift-ci please test source compatibility

hamishknight commented:
@swift-ci please test compiler performance

hamishknight commented:
https://github.com/apple/llvm-project/pull/892

@swift-ci please test macOS

hamishknight commented:
@swift-ci please test compiler performance

hamishknight commented:
@swift-ci please test source compatibility

hamishknight commented:
@swift-ci please test compiler performance

hamishknight commented:
@swift-ci please test source compatibility

hamishknight commented:
https://github.com/apple/llvm-project/pull/892

@swift-ci please test

hamishknight commented:
https://github.com/apple/llvm-project/pull/892

@swift-ci please test

hamishknight commented:
@swift-ci please test source compatibility

hamishknight commented:
@swift-ci please test source compatibility

hamishknight commented:
Ping @CodaFi @slavapestov

hamishknight commented:
@swift-ci please test source compatibility Debug

CodaFi commented on lib/Sema/TypeCheckDecl.cpp:
```diff
@@ -1910,8 +1886,7 @@ FunctionOperatorRequest::evaluate(Evaluator &evaluator, FuncDecl *FD) const {
   }
-
-  return op;
+  return op.get();
```
It's not evident in the diff, but we check this pointer right?

hamishknight replied:
Yup, there's an `if (!op)` statement just above this that emits a diagnostic and bails.

Thanks for the review!

CodaFi commented on lib/AST/NameLookup.cpp:
```diff
@@ -2363,6 +2400,222 @@ swift::getDirectlyInheritedNominalTypeDecls(
 }
 
+template <typename T>
+static TinyPtrVector<T *> lookupOperatorImpl(
```
I wonder if it would be cleaner to factor operator and precedencegroup lookup into its own translation unit. What do you think?

hamishknight replied:
Hmm, could do – in fact it could be argued that it belongs in `ModuleNameLookup.cpp`, what do you think?

CodaFi replied:
The lookup logic has been specialized enough that I think it's become a service in its own right.

CodaFi replied:
You totally don't have to do this now.

hamishknight replied:
@CodaFi I've gone ahead and split it into it's own file, how does it look?

CodaFi commented on lib/AST/NameLookup.cpp:
```diff
@@ -594,11 +595,25 @@ static void recordShadowedDecls(ArrayRef<ValueDecl *> decls,
+}
+
+template <typename T, typename Container>
+static bool removeShadowedDeclsImpl(Container &decls, const DeclContext *dc) {
```
`const` please

hamishknight replied:
For `decls`? The function mutates the passed vector, so would need to be changed to return a new vector, is this what you meant?

CodaFi replied:
Ooh, yikes. Okay, that makes sense.

hamishknight commented:
@swift-ci please test source compatibility Debug

hamishknight committed:
Allow shadowing checks to deal with operator decls

Add overloads of `removeShadowedDecls` that deal
with operator and precedencegroup decls, and
template the existing shadowing logic such that it
can process them.

hamishknight committed:
[Frontend] Add flag to enable new operator lookup

For now disabled by default, but once we get a new
language mode, we should flip it on by default.

hamishknight committed:
[Serialization] Use direct operator lookup for xrefs

We don't need to look at re-exports when resolving
cross references. Luckily the old lookup logic
didn't, but the new logic will. Therefore switch
it over to calling the appropriate request for a
direct operator lookup. In addition, return a
deserialization error instead of silently
returning nullptr if the lookup fails.

hamishknight committed:
Fix precedencegroup and operator decl lookup

Re-implement operator and precedencegroup decl
lookup to use `namelookup::getAllImports` and
existing decl shadowing logic. This allows us to
find operator decls through `@_exported` imports,
prefer operator decls defined in the same module
over imported decls, and fixes a couple of other
subtle issues.

Because this new implementation is technically
source breaking, as we can find multiple results
where we used to only find one result, it's placed
behind the new Frontend flag
`-enable-new-operator-lookup` (with the aim of
enabling it by default when we get a new language
mode).

However the new logic will always be used if the
result is unambiguous. This means that e.g
`@_exported` operators will be instantly available
as long as there's only one candidate. If multiple
candidates are found, we fall back to the old
logic.

Resolves SR-12132.
Resolves rdar://59198796.

hamishknight committed:
[Sema] Diagnose cross-file operator redeclarations

Previously we permitted operator redeclarations
across files as the user could shadow imported
operator decls on a per-file basis, and we would
prefer an imported operator decl over one in
another file.

However with the new operator lookup
implementation, operators in the current module
always shadow imported decls. As such, we should
start diagnosing cross-file redeclarations when
the new lookup logic is enabled.

Resolves rdar://48731166

hamishknight commented:
https://github.com/apple/llvm-project/pull/892

@swift-ci please test

hamishknight commented:
https://github.com/apple/llvm-project/pull/892

@swift-ci please test

hamishknight commented:
https://github.com/apple/llvm-project/pull/892

@swift-ci please test Windows

hamishknight commented:
https://github.com/apple/llvm-project/pull/892

@swift-ci please test Linux

hamishknight commented:
https://github.com/apple/llvm-project/pull/892

@swift-ci please test Linux

hamishknight commented:
https://github.com/apple/llvm-project/pull/892

@swift-ci please smoke test

hamishknight commented:
https://github.com/apple/llvm-project/pull/892

@swift-ci please test

hamishknight commented:
@swift-ci please test source compatibility

hamishknight commented:
@swift-ci please test source compatibility
